### PR TITLE
Add new service builder codegen

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

https://github.com/awslabs/smithy-rs/pull/1679

## Description

- Add `ServerProtocol` interface to allow for server side protocol specific methods.
- Make public the structs merged in https://github.com/awslabs/smithy-rs/pull/1679.
- Add `ServerOperationGenerator`, which generates a ZST and implements `OperationShape` on it.
- Add `ServerServiceGeneratorV2`, which generates the service newtype around a router and a service builder.
- Add `hidden` argument to `RustModule` which allows modules to be marked with `#[doc(hidden)]`.
- Add `BuildModifier` trait to provide a common interface for extending service builders.
- Add `Upgradable` trait to simplifying bounds when upgrading from an `Operation` to a HTTP service.
- Add `FromRequest`, `FromParts`, and `IntoResponse` implementations.
- Make `RoutingService` accept general body types `B` for the inner services `http::Response<B>`.
- Use new service builder in protocol tests.

## Notes

- All new generated code is placed in `operation_shape.rs` and `service.rs`. These are both marked as `#[doc(hidden)]` for now.
- The generated diff should be examined extensively. 
- Follow up tasks:
    - https://github.com/awslabs/smithy-rs/issues/1707
    - https://github.com/awslabs/smithy-rs/issues/1706
    - https://github.com/awslabs/smithy-rs/issues/1705
    - https://github.com/awslabs/smithy-rs/issues/1703

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:64]:
```suggestion
        // Upgrades the core protocol to a `ServerProtocol`.
        fun fromCoreProtocol(coreCodegenContext: CoreCodegenContext, protocol: Protocol): ServerProtocol =
            when (protocol) {
                is AwsJson -> ServerAwsJsonProtocol(coreCodegenContext, protocol)
                is RestJson -> ServerRestJsonProtocol(coreCodegenContext, protocol)
                is RestXml -> ServerRestXmlProtocol(coreCodegenContext, protocol)
                else -> throw IllegalStateException("unsupported protocol: $protocol")
            }
```

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:93]: Use `ServerCargoDependency.SmithyHttpServer(runtimeConfig).asType().namespace` here and in the other protocols.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:69]: So are these meant to replace the existing server protocol classes? Like, the `ServerAwsJson` class in `ServerAwsJson.kt` etc.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:98]:
```suggestion
    override fun routerType() = RuntimeType("AwsJsonRouter", ...
```

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:103]: I guess it's because we never had a need to generate a Rust type for a `service` shape. Now only the server does. We should track these in an issue to work on creating our own bottom-most symbol provider, tailored for the server codegen needs, now that we're starting to see how they need to deviate.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:109]: You can inline this where it's used and remove the `.dq()`; you can use double-quotes inside triple double-quotes.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:126]:
```suggestion
            "Router" to routerType(),
            "Pairs" to pairs,
```

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:69]: I prefer this composition approach over the inheritance one we used over there though. Can you them into these ones, and split in several files?

And the factories should now return these ones e.g. `ServerRestJsonFactory`'s `protocol` method should return `ServerRestJson` etc...

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:98]: You can define one-line functions without creating a `{}` block like this. There are several places in the diff, especially in this file, where you could use this to reduce verbosity a bit.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:212]: nit: here and in line 223, these are empty lines, when we want them to be indented 12 spaces, so that the indent of the until the start of the block is constant.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:222]: I guess we should define `Pin`, `Box`, `Future` in `RuntimeTypes.kt`

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/RustModule.kt
Comment (david-perez) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/RustModule.kt:8]: Whether the module is doc hidden or not should be indicated via the input `RustMetadata`. Example:

```
RustMetadata(additionalAttributes = listOf(Attribute.DocHidden))
```

And then line 14 takes care of rendering the doc hidden attribute.

File: rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs:55]: I think it's best we don't modify the public Pokémon Service example until we're ready to move people to the new builder and deprecate the old one.

We should instead create a Kotlin unit test that generates a simple model and uses the new builder instead. It could also assert that the new builder is generated as doc hidden in `lib.rs`.

File: rust-runtime/aws-smithy-http-server/src/build_modifier.rs
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/build_modifier.rs:8]: Missing Rust docs?

File: rust-runtime/aws-smithy-http-server/src/runtime_error.rs
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/runtime_error.rs:60]: Nice. Is there any blocker now to making `RuntimeError` generic over the protocol, and implement protocol-specific implementations of `IntoResponse` for them? It's not blocked on moving to the new builder right?

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:69]: I'm not sure how to best merge them since inheritance has the advantage that we just override the method of the `Protocol` class that needs to return a different generator or that is subtly difference, and everything works correctly.

Maybe we need to use a combination of both: the truly server-specific methods go into this `ServerProtocol` interface, but then the classes that implement this interface should also inherit from `Protocol` to override things like e.g. `structuredDataSerializer`?

**Review by david-perez — COMMENTED** [review]
Codegen looks good, just a few nits.

I guess the main two things to merge this are:
* A unit test instead of modifying the Pokémon service to use the new service builder; and
* Figuring out how best to merge the server protocol classes so we don't end up with duplicates, one implementing the `ServerProtocol` interface this PR introduces, and the others inheriting from `Protocol`...

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt:87]: This was done in anticipation of having multiple services generated at the same time, in which case `CodegenContext` would not store a single `serviceShape`. As I understand it now, this is not really a goal. I'll revert it.

File: rust-runtime/aws-smithy-http-server/src/runtime_error.rs
Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/runtime_error.rs:60]: After this is merged I think we could replace

```rust
#[derive(Debug)]
pub struct RuntimeError {
    pub protocol: Protocol,
    pub kind: RuntimeErrorKind,
}

impl<P> IntoResponse<P> for RuntimeError {
    fn into_response(self) -> http::Response<crate::body::BoxBody> {
        self.into_response()
    }
}

impl RuntimeError {
    pub fn into_response(self) -> Response {
        /* runtime match on enum */
```

with

```rust
#[derive(Debug)]
pub struct RuntimeError {
    pub kind: RuntimeErrorKind,
}

impl IntoResponse<RestJson1> for RuntimeError {
    fn into_response(self) -> http::Response<crate::body::BoxBody> {
        /* protocol specific implementation */
    }
}
```

I think this should still be superseded by https://github.com/awslabs/smithy-rs/issues/1703 where we do away with `RuntimeError` completely and have protocol specific enums and impls.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:222]: Side note: I hope that this associated type will not exist as a `Pin<Box<_>>` for long - when we are landing https://github.com/awslabs/smithy-rs/issues/1706 we should be able to give this a concrete type.

File: rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs
Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs:56]: I think this would give an error saying that `Op` and `Exts` are unconstrained.

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs:56]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d42bc87814ba44f09e5df4464f084fbd

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs:56]: This is a bit closer but `Ext` is still unconstrained unfortunately.

```rust
impl<Op, H, Exts> From<H> for Operation<IntoService<Op, H>>
where
    Op: OperationShape,
    H: Handler<Op, Exts>,
{
    fn from(h: H) -> Self {
        todo!()
    }
}
```

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/main.rs:56]: If `Exts` was an associated type on `Handler` then this would work and worth experimenting with. It might have some knock on effects elsewhere though.

**Review by Velfi — APPROVED** [review]
I only looked at the changes that affect the SDK. LGTM 😅

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:333]: I've just inlined the alias. We have `type BoxError` in nearly all the runtime crates - it feels odd to me to `pub` when the others aren't.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:89]: I've extended

```kotlin
    fun Protocol(runtimeConfig: RuntimeConfig) =
        RuntimeType("Protocol", ServerCargoDependency.SmithyHttpServer(runtimeConfig), "${runtimeConfig.crateSrcPrefix}_http_server::protocols")

```
in `codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ServerRuntimeType.kt` with
```kotlin
    fun Protocol(name: String, runtimeConfig: RuntimeConfig) =
        RuntimeType(name, ServerCargoDependency.SmithyHttpServer(runtimeConfig), "${runtimeConfig.crateSrcPrefix}_http_server::protocols")

    fun Protocol(runtimeConfig: RuntimeConfig) = Protocol("Protocol", runtimeConfig)
```

Unresolve if you're not happy with this.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:47]: I had a fleeting thought that perhaps there should be some sort of associated type `RoutingInfo` (which is `String` for `AwsJsonRouter` and `RequestSpec` for `RestRouter`.) on some trait which provides a constructor.

I don't think this gives us any benefits though - I'm going to delete this TODO.

File: rust-runtime/aws-smithy-http-server/src/build_modifier.rs
Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/build_modifier.rs:8]: Removed this completely, will add it back in a corollary PR.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:64]: Will track this effort/investigation here https://github.com/awslabs/smithy-rs/issues/1698

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt:86]: https://github.com/awslabs/smithy-rs/issues/1724

**Review by david-perez — APPROVED** [review]

**Comment by hlbarber** [comment]
Closes https://github.com/awslabs/smithy-rs/issues/1705

**Pull request merged**
